suite/tests/client-interface/tls.dll.cpp

```diff
@@ -196,24 +197,37 @@ test_calloc(void)
             if (dr_raw_tls_calloc(&reg, &offset, num_slots, alignment)) {
                 ASSERT(alignment == 0 || ALIGNED(offset, alignment));
                 total_allocated += num_slots;
-                if (offset < min_offset)
-                    min_offset = offset;
+                offsets[num_allocs] = offset;
+                sizes[num_allocs] = num_slots;
+                ++num_allocs;
+                if (num_allocs == max_allocs)
+                    break;
             } else {
                 if (num_slots == 1 && alignment <= 1)
                     break;
             }
         }
         ASSERT(total_allocated > 0);
-        if (previously_allocated == 0)
-            previously_allocated = total_allocated;
-        else
-            ASSERT(previously_allocated == total_allocated);
+        /* If space was exhausted, total allocated should be the same each time. */
+        if (num_allocs < max_allocs) {
+            if (previously_allocated == 0)
+                previously_allocated = total_allocated;
+            else
+                ASSERT(previously_allocated == total_allocated);
+        }
         /* Free the slots. */
-        ASSERT(dr_raw_tls_cfree(min_offset, total_allocated));
-        /* Check that a second attempt to free fails. */
-        ASSERT(!dr_raw_tls_cfree(min_offset + rand() % total_allocated, 1));
-    }
+        for (size_t i = 0; i < num_allocs; i++) {
+            bool try1 = dr_raw_tls_cfree(offsets[i], sizes[i]);
+            ASSERT(try1);
+#ifndef WINDOWS
+            /* FIXME: On Unix a double free just returns false, but on Windows
```

Comment:
Please avoid FIXME in new code unless it is a significant missing piece.  We also always put in an issue number for anything serious that deserves a TODO or FIXME: so do not add a TODO (or FIXME) without an issue number.  But I don't think this deserves either.   Maybe this could be `XXX` (for which we still prefer an issue number) or just a plain comment as presumably release build Windows is a nop for double free so it's already gracefully handled (and the debug assert does match other DR API routines which assert on fatal invalid parameters (like nulls) in debug build...maybe would do things differently if starting over now but that's the current API convention) -- though arguably this is not fatal; but there is the precedent in other allocators like libc free() of not returning nice error messages on double free which somehow feels like a different kind of user error....though maybe it shouldn't).